### PR TITLE
feat: improve output of ProjectGraphTask and include topological sort.

### PR DESCRIPTION
**Pull request by autonomousapps**
In a follow-up, I will add a task at the root level for doing this globally in one pass.

### CODE REVIEW

File: src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt [hunk 1]
```diff
@@ -38,4 +44,29 @@ internal object GraphWriter {
...
+   */
+  fun topological(graph: Graph<Coordinates>): String {
+    return Topological(graph, graph.root())
+      .order
+      .joinToString(separator = "\n") {
+        // ":foo:bar 0"
+        // ":other:baz 1"
+        it.maybeProjectCoordinates(buildPath).gav() + " ${graph.inDegree(it)}"
+      }
+  }
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt:63]: @staktrace this is the main area of interest for us currently.

I can think of very expensive algorithms that check if there's a path from one node to another, and using that to create the batches, but I feel like there might also be some common(ish) graph algo that can also cut up a graph into mutually semi-independent chunks. I'll page through the lit, but if you can think of anything I'm all ears.

Comment (autonomousapps) [reply src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt:63]: Omg, I just realized that what I want is basically just the dominance tree. It already contains all the information necessary to go in order in a maximally concurrent way. At least, I think. Will test later!

Comment (staktrace) [reply src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt:63]: Yeah I think using the dominator tree of the dependency graph makes sense here.

File: graph-support/src/main/kotlin/com/autonomousapps/graph/Graphs.kt
Comment (autonomousapps) [inline graph-support/src/main/kotlin/com/autonomousapps/graph/Graphs.kt:1]: Need to publish a new version of graph-support before I can merge.

File: src/functionalTest/groovy/com/autonomousapps/jvm/WorkPlanSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.autonomousapps.jvm.projects.WorkPlanProject
+import spock.lang.PendingFeature
+
+import static com.autonomousapps.utils.Runner.build
+import static com.google.common.truth.Truth.assertThat
+
+final class WorkPlanSpec extends AbstractJvmSpec {
+
+  @PendingFeature
```
Comment (autonomousapps) [inline src/functionalTest/groovy/com/autonomousapps/jvm/WorkPlanSpec.groovy:11]: Fix test or delete feature and test.

File: src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt [hunk 3]
```diff
@@ -38,4 +45,56 @@ internal object GraphWriter {
...
+      top.forEach { node ->
+        appendLine(node.maybeProjectCoordinates(buildPath).gav())
+        // append(" ")
+        // appendLine("${paths.distanceTo(node)}")
+      }
+    }
+  }
+
+  // TODO replace with Hu's algorithm (for scheduling concurrent work), and push relevant bits into Graphs.kt.
+  fun workPlan(graph: Graph<Coordinates>): String {
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/graph/GraphWriter.kt:69]: Follow-up.

**Pull request merged**